### ISSUE TITLE
False positive for `Lint/UselessAssignment` when assigning in block

### ISSUE THREAD

**Issue by Earlopain** [rubocop/rubocop#14457]
Also see https://github.com/rubocop/rubocop/issues/14445#issuecomment-3194813998 (this shows an offense at a different location)

## Expected behavior

No offense for the following code:

```rb
def foo
  node = self
  node.children&.pop&.each do |node|
    result = node if node.has_children?
  end
  result
end
```

## Actual behavior

```
test.rb:4:5: W: [Correctable] Lint/UselessAssignment: Useless assignment to variable - result.
    result = node if node.has_children?
    ^^^^^^
```

## RuboCop version

Include the output of `rubocop -V` or `bundle exec rubocop -V` if using Bundler.
If you see extension cop versions (e.g. `rubocop-performance`, `rubocop-rspec`, and others)
output by `rubocop -V`, include them as well. Here's an example:

```
$ [bundle exec] rubocop -V
1.79.2 (using Parser 3.3.9.0, rubocop-ast 1.46.0, analyzing as Ruby 2.7, running on ruby 3.4.3) [x86_64-linux]
  - rubocop-performance 1.25.0
  - rubocop-rake 0.7.1
  - rubocop-rspec 3.6.0
```

**Comment by Earlopain** [comment]
Nevermind, this is working as intended. My bad. The autocorrect for `Style/For` is unsafe because of scope changes like this.
